README.md
[![Build Status Travis](https://travis-ci.org/terhnt/unowallet.svg?branch=develop)](https://travis-ci.org/terhnt/unowallet)
[![Build Status Circle](https://circleci.com/gh/terhnt/unowallet.svg?&style=shield)](https://circleci.com/gh/terhnt/unowallet)
[![Slack Status](http://slack.counterparty.io/badge.svg)](http://slack.counterparty.io)

Unowallet
================

Online Webwallet for [Unoparty](http://www.unoparty.io).

Originally based off of [Carbonwallet](http://www.carbonwallet.com) (however virtually all the original code has been removed or rewritten).


Production Systems
-------------------

* Main page: **[unowallet.io](https://unowallet.io/)**
* Primary server -- Mainnet: **[wallet.unowallet.io](https://wallet.unowallet.io/)**
* Primary server -- Testnet: **[testnet.unowallet.io](https://testnet.unowallet.io/)**


Features
----------

- Deterministic wallet addresses (BIP 32-based)
- Supports the majority of Unoparty functionality
- Fully-AJAX driven
- Anonymous
- Runs in the browser, with keys created in memory
- Multi-sig

Browser Support
-------------------

**Desktop**

- Chrome 23+ (preferred browser)
- Firefox 25+
- Safari 7+
- Opera 15+

Notably, Internet Explorer is **not** supported, due to its lack of full Content-Security-Policy support (even with IE 11).

**Mobile**

- IOS Safari 7+
- Android Browser 4.4+
- Chrome for Android 33+
- Chrome for iOS 35+
- Firefox for Android 26+


Build Instructions
-------------------

### Before running the build system:
```
sudo npm install -g grunt-cli bower browserify
```

### To build:
```
cd src
bower install
cd vendors/bitcoinjs-lib
npm install
browserify --standalone bitcoinjs src/index.js | uglifyjs -c --mangle reserved=['BigInteger','ECPair','Point'] -o bitcoinjs.min.js
cd ../../../
npm install
```

### To (re)build the static (i.e. minified) site:
```
grunt build
```

### To regenerate dependencies hash file (```src/.bowerhashes```):
```
grunt freeze
```

### To enable localizations (optional):
1. Create an account on [Transifex](https://www.transifex.com/)
2. In your home directory, create a file named `.transifex` and put your Transifex username and password into it in this format: `user:password`
3. Run `grunt build` to download translations
4. Add the languages you want to support to `AVAILABLE_LANGUAGES` in **unowallet.conf.json** - you can use **unowallet.conf.json.example** as a template. The template file contains **only** the setting relevant to languages and does not replace the rest of variables required in that file (refer to Federeated Node documentation for additional details about `unowallet.conf`).

Setting up your own Unowallet Server
-----------------------------------------

See [this link](http://counterparty.io/docs/federated_node/) for more info.

Development
-----------

The easiest way to develop locally is to install Federated Node in Docker environment. If you already have Unobtanium Core addrindex, Unoparty Server and Unoblock, then Unowallet can be manually installed using either of these approaches:

* Stand-alone Docker environment: refer to docker\start.sh in the Unowallet source code.
* Manually: example based on the Docker template for Unowallet on Ubuntu 16.04 is provided below.

### Manual installation in local environment (Ubuntu 16.04):
1. Clone repository and `cd` into result folder
2. NodeJS 9.x + NPM installation (if you don't have it):
```
curl -sL https://deb.nodesource.com/setup_9.x | sudo -E bash -
sudo apt-get install -y nodejs
```
5. Dowload dependencies in the root folder: `npm install`
6. Build: `npm run build`. The application will be stored in the `./build` subdirectory.
4. Configure: Copy `unowallet.conf.json.example` to .`/build/unowallet.conf.json`.
  - Notice the `servers` parameter in JSON file. It's responsible for API endpoint(s). By default it's `https://wallet.unowallet.io`
6. Run: `npm run serve`
7. Use: Visit `http://localhost:3000`

#### Notes:

* the `--dontcheckdeps` speeds up the process and avoids having to do `grunt freeze` everytime you make a change to a dependency during development
* If you want to test your local version on another device (or let another person test something) use https://ngrok.com to setup a tunnel to your local environment
* If you want to use HTTPS, refer to additional steps required in the Unowallet Docker start script

### Note concerning `npm install`
`npm install` triggers a `prepublish` which is configured to do `grunt build`
and will bork if you haven't done a `grunt freeze` after making changes to dependencies.
You can use `npm update` to circumvent this during development.

### Running tests in browser
You can run tests in your browser by doing the above steps and;
 - open a seperate terminal and [from the root of the project, not from `build/` run `serve -p 3001` (different port)
 - visit `http://localhost:3001/test/test.html`

### Running tests from CLI (using phantomjs headless browser)
 - `npm test`

### Development without a full node

To work on unowallet without running a full node and supporting services locally, please see the scripts located in the [local development folder](local-development).


License
-------------------

http://opensource.org/licenses/CDDL-1.0


CHANGELOG.md
### v1.8.0 ###
**Release Date:** June 26, 2016

**Enhancements:**
- Docker image builds
- Eliminate socket.io feed code (which was causing issues/not working), replace with simple polling

**Fixes:**
- Fix exchange market display -- is now sortable (#768)
- UI tweaks (#780)
- Disable portfolio and leaderboard, clean up menu interface
- Update bitcore to newest
- Fix date field on broadcast modal (#779)
- Remove asset description limit
- Fix DEX buy/sell validation error messages (#746)
- Better error handling (pops up an error dialog when many times before it would just error to console)
- Pull up to 5000 history entries instead of 100 (fixes #565)
- Fix minor display issues with issuances (#551, #427)

### v1.7.0 ###
**Release Date:** January 24, 2016

**Enhancements:**
- Update to work with newest `bitcore` and enforced high S signatures (@rubensayshi)
- Extensive code cleanup/reformatting (@rubensayshi)
- Added basic Travis and CircleCI test suite

**Removed features:**
- Remove use of `counterwebdeps` external repo.

**Fixes:**
- Fixed improper disabled textarea on Firefox in signing raw transactions dialog

### v1.6.3 ###
**Enhancements:**
- Allow `OP_RETURN`; use`'auto'` encoding

**Removed features:**
- Chat

**Fixes:**
- Fix `refreshUnopartyBalances()`
- Fix pending message for issuance with quantity=0

### v1.6.2 ###
**Fixes:**
- Proper formatting for multisig addresses on history page and stats page
- Updated terminology in several parts of the UI
- Updated blockscan links in the history page and stats page
- Blockscan links now work with testnet
- A few other minor changes

### v1.6.1 ###
**Fixes:**
* Bug fixes: http://goo.gl/uj5mcQ

### v1.6.0 ###
**Enhancements:**
- Full multisig support
- Free numeric asset and dividend fee support
- Better internationalization
- Make Unowallet possible to package in Linux

**Removed features:**
- BTC Trading
- Callback
- RPS

**Fixes:**
* Numerous fixes: http://goo.gl/JUPoBp
* Minor UI tweaks

Full list of changes are at: http://goo.gl/V3y29V

### v1.5.0 ###
**Enhancements:**
* Added Quick Buy interface with Vennd Machine
* Added internationalization support
* New address generation system
* DEX design improvement
* Added order expiration and fees in advanced option
* Added quick url access with choosen password

**Fixes:**
* Numerous fixes: http://goo.gl/opzkoh
* Minor UI tweaks

Full list of changes are at: http://goo.gl/AICdol

### v1.4.0 ###
**Enhancements:**
* Added new DEx interface
* Added Armory Offline signing support
* Added page-level help for balances and exchanges (with more coming in the future, as necessary)

**Fixes:**
* Significant performance and reliability improvements (on the unoblockd side)
* Numerous minor fixes
* Minor UI tweaks

### v1.3.0 (~2014-07-13) ###
**Enhancements:**
* Added Rock-Paper-Scissors
* Added Zeroconf support (show pending balances for assets)
* Added IP-based blocking for betting, RPS and dividend issuance. We are currently blocking people
  with US-based IP addresses from using these features on unowallet.io itself.

**Fixes:**
* Fixed issue with asset name validation (length checking was incorrect)
* Fixed issue with History page not showing balance history properly anymore
* Fixed up and improved stats page
* Numerous minor fixes

### v1.2.2 (~2014-06-25) ###
**Enhancements:**
* Check double connection to prevent double btcpay
* Cancel bet buttons

### v1.2.1 ###
**Enhancements:**
* Added wallet-level statistics and logging of wallet logins

**Fixes:**
* Broadcast feeds with enhanced feed info with a protocol prefix now are seen properly


### v1.2.0 ###
**Enhancements:**
* Added in binary option and CFD functionality
* Simplified menu layout

**Fixes:**
* numerous minor fixes

### v1.1.6 ###
**Enhancements:**
* Phrasing change: Dividends -> Distributions, Asset -> Token

**Fixes:**
* numerous minor fixes

### v1.1.5 ###
**Enhancements:**
* moved to using the bitcore library

**Fixes:**
* numerous minor fixes

### v1.1.4 ###
**Enhancements:**
* added stats page

**Fixes:**
* ignore dust with orders
* open orders in the open order box should decrease properly as they are matched
* moved to using VWAP over last 8 trades for the market price

### v1.1.0 ###
**Enhancements:**
* A restrictive content security policy (CSP) is in use, which GREATLY reduces the exposure and effectiveness of any cross site scripting (XSS) attacks. Inline script injection/eval is basically impossible now (unless it is a browser bug, or being done by an installed browser plugin). This required a code overhaul to support. More info: http://www.html5rocks.com/en/tutorials/security/content-security-policy/

**Fixes:**
* Tightened up XSS defenses numerous places
* Wallet passphrase generation algorithm improved (added entropy and removed fallback to the JS-based prng, which had issues)
* Numerous other security improvements
* Sweeping overhauled...should work much better now
* Many smaller bug fixes

### v1.0.0 ###
* MAINNET RELEASE
* numerous bug fixes

### v0.9.5 ###
**Enhancements:**
* implemented asset description json and asset pictures, yay (see https://wiki.counterparty.co/w/Enhanced_Asset_Info_in_Counterwallet)
* added 7 sample moving average to price charts
* show # characters remaining when typing asset description
* we now wait 6 blocks on an order match before doing a BTCpay (we do this for protection against a reorg making an order match disappear)

**Fixes:**
* open trades show properly in non-BTC pairs (thanks rotalumis for pointing this out)
* debugging info now shows block number and last message ID properly, as well as backend API URL list
* go off of tx_hash instead of tx_index where it matters, as tx_index can change between updates
* fixed bug in order fee_required/fee_provided dispay (was showing the full amount instead of the remaining amount)
* when overriding the market rate, clear the text box when switching between "as unit price" and "as quantity" (thanks rotalumis)

src/js/consts.js
/***********
 * GLOBAL CONSTANTS
 ***********/
var VERSION = "1.9.0";
var PREFERENCES = {}; //set when logging in

//Addresses
var DEFAULT_NUM_ADDRESSES = 1; //default number of addresses to generate. Go with 1 for now to be more newbie friendly
var MAX_ADDRESSES = 20; //arbitrary (but will generate more on login if they have activity...this just prevents
                        //additional addresses from being generated via the GUI)

//Order expiration
var ORDER_DEFAULT_EXPIRATION = 8064; //num blocks until expiration (~2 months)
var ORDER_MAX_EXPIRATION = 8064; //max expiration for order

var STATS_MAX_NUM_TRANSACTIONS = 100; //max # transactions to show in the table
var VIEW_PRICES_NUM_ASSET_PAIRS = 50; //show market info for this many pairs
var VIEW_PRICES_ASSET_PAIRS_REFRESH_EVERY = 5 * 60 * 1000; //refresh asset pair market info every 5 minutes
var VIEW_PRICES_NUM_LATEST_TRADES = 50; //show this many latest trades on the view prices page
var VIEW_PRICES_LATEST_TRADES_REFRESH_EVERY = 5 * 60 * 1000; //refresh latest trades every 5 minutes

var MARKET_INFO_REFRESH_EVERY = 5 * 60 * 1000; //refresh market info every 5 minutes while enabled (on buy/sell page, and view prices page)

var CHAT_NUM_USERS_ONLINE_REFRESH_EVERY = 5 * 60 * 1000; //refresh online user count every 5 minutes while enabled

var ALLOW_UNCONFIRMED_INPUTS = true;  // allow use unconfirmed unspents

// should be a i18n key
var ACTION_PENDING_NOTICE = "pending_notice";

var ARMORY_OFFLINE_TX_PREFIX = "=====TXSIGCOLLECT-";

var DEFAULT_PREFERENCES = {
  'num_addresses_used': DEFAULT_NUM_ADDRESSES,
  'num_segwit_addresses_used': DEFAULT_NUM_ADDRESSES,
  'address_aliases': {},
  'selected_theme': 'ultraLight',
  'selected_lang': 'en-us',
  'watch_only_addresses': [],
  'armory_offline_addresses': [],
  'multisig_addresses': [],
  'has_accepted_license': false
};

var UNOWALLET_CONF_LOCATION = "/unowallet.conf.json";

var NUMERIC_ASSET_ID_MIN = bigInt(26).pow(12).add(1);
var NUMERIC_ASSET_ID_MAX = bigInt(256).pow(8);

var SUBASSET_MAX_DISP_LENGTH = 20;

var IS_MOBILE_OR_TABLET = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
var MAX_INT = Math.pow(2, 63) - 1;
var UNIT = 100000000; //# satoshis in whole
var MIN_FEE = 20000; // in satoshis (== .0002 BTC)
var REGULAR_DUST_SIZE = 5430;
var MULTISIG_DUST_SIZE = 7800;
var MIN_BALANCE_FOR_ACTION = 50000; //in satoshis ... == .0005
var ASSET_CREATION_FEE_XCP = 0.5; //in normalized XCP
var SUBASSET_CREATION_FEE_XCP = 0.25; //in normalized XCP
var DIVIDEND_FEE_PER_HOLDER = 0.0002
var MAX_ASSET_DESC_LENGTH = 41; //42, minus a null term character?
var FEE_FRACTION_REQUIRED_DEFAULT_PCT = .9;   //0.90% of total order
var FEE_FRACTION_PROVIDED_DEFAULT_PCT = 1;   //1.00% of total order
var FEE_FRACTION_DEFAULT_FILTER = .95;
var BTC_ORDER_MIN_AMOUNT = 0.01;
var B26_DIGITS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
var ORIG_REFERER = document.referrer;

var ENTITY_NAMES = {
  'burns': 'Burn',
  'melts': 'Melt',
  'debits': 'Debit',
  'credits': 'Credit',
  'sends': 'Send',
  'orders': 'Order',
  'order_matches': 'Order Match',
  'btcpays': 'BTCPay',
  'issuances': 'Issuance',
  'broadcasts': 'Broadcast',
  'bets': 'Bet',
  'bet_matches': 'Bet Match',
  'dividends': 'Distribution',
  'cancels': 'Cancel',
  'callbacks': 'Callback',
  'bet_expirations': 'Bet Expired',
  'order_expirations': 'Order Expired',
  'bet_match_expirations': 'Bet Match Exp',
  'order_match_expirations': 'Order Match Exp',
  'rps': 'Rock-Paper-Scissors',
  'rps_matches': 'RPS Match',
  'rpsresolves': 'RPS Confirmed',
  'rps_expirations': 'RPS Expired',
  'rps_match_expirations': 'RPS Match Expired'
};

var ENTITY_ICONS = {
  'burns': 'fa-fire',
  'melts': 'fa-fire',
  'debits': 'fa-minus',
  'credits': 'fa-plus',
  'sends': 'fa-share',
  'orders': 'fa-bar-chart-o',
  'order_matches': 'fa-exchange',
  'btcpays': 'fa-btc',
  'issuances': 'fa-magic',
  'broadcasts': 'fa-rss',
  'bets': 'fa-bullseye',
  'bet_matches': 'fa-exchange',
  'dividends': 'fa-ticket',
  'cancels': 'fa-times',
  'callbacks': 'fa-retweet',
  'bet_expirations': 'fa-clock-o',
  'order_expirations': 'fa-clock-o',
  'bet_match_expirations': 'fa-clock-o',
  'order_match_expirations': 'fa-clock-o',
  'rps': 'fa-trophy',
  'rps_matches': 'fa-trophy',
  'rpsresolves': 'fa-trophy',
  'rps_expirations': 'fa-trophy',
  'rps_match_expirations': 'fa-trophy'
};

var ENTITY_NOTO_COLORS = {
  'burns': 'bg-color-yellow',
  'melts': 'bg-color-yellow',
  'debits': 'bg-color-red',
  'credits': 'bg-color-green',
  'sends': 'bg-color-orangeDark',
  'orders': 'bg-color-blue',
  'order_matches': 'bg-color-blueLight',
  'btcpays': 'bg-color-orange',
  'issuances': 'bg-color-pinkDark',
  'broadcasts': 'bg-color-magenta',
  'bets': 'bg-color-teal',
  'bet_matches': 'bg-color-teal',
  'dividends': 'bg-color-pink',
  'cancels': 'bg-color-red',
  'callbacks': 'bg-color-pink',
  'bet_expirations': 'bg-color-grayDark',
  'order_expirations': 'bg-color-grayDark',
  'bet_match_expirations': 'bg-color-grayDark',
  'order_match_expirations': 'bg-color-grayDark',
  'rps': 'bg-color-blue',
  'rps_matches': 'bg-color-blueLight',
  'rpsresolves': 'bg-color-blue',
  'rps_expirations': 'bg-color-blueLight',
  'rps_match_expirations': 'bg-color-blueLight'
};

var BET_TYPES = {
  0: "Bullish CFD",
  1: "Bearish CFD",
  2: "Equal",
  3: "Not Equal"
};

var BET_TYPES_SHORT = {
  0: "BullCFD",
  1: "BearCFD",
  2: "Equal",
  3: "NotEqual"
}

var BET_TYPES_ID = {
  "BullCFD": 0,
  "BearCFD": 1,
  "Equal": 2,
  "NotEqual": 3
}

var COUNTER_BET = {
  "Equal": 3,
  "NotEqual": 2,
  "BullCFD": 1,
  "BearCFD": 0
}

var BET_MATCHES_STATUS = {
  "settled: liquidated for bear": 0,
  "settled: liquidated for bull": 1,
  "settled: for equal": 2,
  "settled: for notequal": 3
}

var LEVERAGE_UNIT = 5040;

var MAINNET_UNSPENDABLE = 'uNopartyXburnXXXXXXXXXXXXXXXWJmsqn';
var TESTNET_UNSPENDABLE = 'UUnoPartyXburnTestnetXXXXXXXXFEeN4';
var REGTEST_UNSPENDABLE = '1UnopartyxBurnxAddrXXXXXXXXXX42dPh';
var TESTNET_BURN_START = 14500;
var TESTNET_BURN_END = 400000;
var REGTEST_BURN_START = 101;
var REGTEST_BURN_END = 150000000;
var MAINNET_BURN_START = 1810000;
var MAINNET_BURN_END = 1810000 + (30*24*60/3);

/***********
 * DYNAMICALLY SET
 ***********/
var TESTNET_PASSPHRASE = qs("passphrase");

var CRYPTED_PASSPHRASE;
if (location.hash.indexOf('cp=') == 1) {
  CRYPTED_PASSPHRASE = location.hash.substr(4);
  location.hash = '';
}
location.hash = '';

//CONSTANTS THAT DEPEND ON IS_DEV / USE_TESTNET
var USER_COUNTRY = ''; //set in logon.js
var CURRENT_PAGE_URL = ''; // set in loadUrl()

//selective disablement
var DISABLED_FEATURES_SUPPORTED = ['betting', 'dividend', 'exchange', 'leaderboard', 'portfolio', 'stats', 'history']; //what can be disabled
var DISABLED_FEATURES = []; //set in unowallet.js

// restricted action
var RESTRICTED_AREA = {} // set in unowallet.js

var RESTRICTED_AREA_MESSAGE = {
  'pages/simplebuy.html': 'buy_xcp_if_legal'
}

var MAX_SUPPORT_CASE_PROBLEM_LEN = 4096;
var QUOTE_ASSETS = []; // initalized with unoblock is_ready()

var QUICK_BUY_ENABLE = false;
var BETTING_ENABLE = true;
var GAMING_ENABLE = true;

function qs(key) {
  //http://stackoverflow.com/a/7732379
  key = key.replace(/[*+?^$.\[\]{}()|\\\/]/g, "\\$&"); // escape RegEx meta chars
  var match = location.search.match(new RegExp("[?&]" + key + "=([^&]+)(&|$)"));
  return match && decodeURIComponent(match[1].replace(/\+/g, " "));
}

//Allow the site root to specify "dev" and "testnet" parameters...
// IS_DEV is enabled if the initial (root) URL access has ?dev=1
// USE_TESTNET is enabled if the initial (root) URL access has ?testnet=1, OR the hostname visited starts with 'testnet' (e.g. testnet.myhost.com)
var IS_DEV = (location.pathname == "/" && qs("dev") && qs("dev") != '0' ? true : false);
var USE_REGTEST = (location.pathname == "/" && qs("regtest") && qs("regtest") != '0' ? true : false);
var USE_TESTNET = (   (((location.pathname == "/" || location.pathname == "/src/" || location.pathname == "/build/") && qs("testnet") && qs("testnet") != '0')
  || location.hostname.indexOf('testnet') != -1) ? true : false
);

var BLOCKEXPLORER_URL = USE_TESTNET ? "https://testnet.xchain.io" : "https://xchain.io";
var GOOGLE_ANALYTICS_UAID = null; //will be set in unowallet.js
var ROLLBAR_ACCESS_TOKEN = null; //will be set in unowallet.js

var TRANSACTION_DELAY = 5000; // delay between transaction to avoid error -22 (vin reused)
var TRANSACTION_MAX_RETRY = 5; // max retry when transaction failed (don't include first transaction, so 3 retry means 4 queries)

var DONATION_ADDRESS = USE_TESTNET ? 'UVKQywFfmP5TvbLsEwQaXTbLspck8UuK3w' : 'uVVuwXm2mDK9pr9XkWT5k7ihQyoSC8y2MW';

var APPROX_SECONDS_PER_BLOCK = USE_TESTNET ? 3 * 60 : 3 * 60; //a *rough* estimate on how many seconds per each block (used for estimating open order time left until expiration, etc)

var KEY_ASSET = {
  'BTC': 'UNO',
  'XCP': 'XUP',
  'USD': 'USD',
  'Bitcoin': 'Unobtanium',
  'Counterparty': 'Unoparty'
};

var KEY_ASSET_WEBSITE = {
  'BTC': 'https://unobtanium.uno',
  'XUP': 'https://unoparty.io/'
};

bitcoinjs.networks.regtest = {
    messagePrefix: '\x1bUnobtanium Signed Message:\n',
    bech32: 'ucrt',
    bip32: {
      public: 0x043587cf,
      private: 0x04358394
    },
    pubKeyHash: 0x0,
    scriptHash: 0x05,
    wif: 0x80
  }

  bitcoinjs.networks.testnet = {
      messagePrefix: '\x1bUnobtanium Signed Message:\n',
      bech32: 'uctn',
      bip32: {
        public: 0x043586CE,
        private: 0x04358293
      },
      pubKeyHash: 0x44, //68
      scriptHash: 0x1E, //30
      wif: 0xEF //239
    }

    bitcoinjs.networks.mainnet = {
        messagePrefix: '\x1bUnobtanium Signed Message:\n',
        bech32: 'ucmn',
        bip32: {
          public: 0x0488B21E,
          private: 0x0488ADE4
        },
        pubKeyHash: 0x82, //130
        scriptHash: 0x1E, //30
        wif: 0xE0 //224
      }

//bitcoinjs.networks.mainnet = bitcoinjs.networks.bitcoin // support for bitcore's name
